lib/vector-core/src/config/mod.rs

```diff
@@ -339,6 +340,31 @@ impl LogNamespace {
         }
     }
 
+    /// Vector: The `ingest_timestamp`, and `source_type` fields are added to "event metadata", nested
+    /// under the name "vector". This data will be marked as read-only in VRL.
+    ///
+    /// Legacy: The values of `source_type_key`, and `timestamp_key` are stored as keys on the event root,
+    /// only if a field with that name doesn't already exist.
+    pub fn insert_standard_vector_source_metadata(
+        &self,
+        log: &mut LogEvent,
+        source_name: &'static str,
+        now: DateTime<Utc>,
```

Comment:
Real nitpick, but calling it `now` is slightly misleading, since it may be the

> S3 object's last modified time

according to a comment below..

Reply:
It _should_ be `now` - it's really incorrect behavior otherwise. We need to determine how we want to resolve, and if that resolution affects the legacy namespacing or just vector namespacing.

Reply:
This _must_ be the current time when using the `Vector` namespace. I'd suggest keeping this parameter but renaming it, and use it for the `Legacy` namespace to keep the current behavior of whatever the source currently does, and for the `Vector` namespace, just get the current time.

Reply:
(If this is a performance concern, or we want to guarantee the timestamp is the same as another place, maybe we could accept 2 different timestamps)

Reply:
The manual way I've handled this for the upcoming `kubernetes_logs` PR has:

```rust
    match (log_namespace, ingestion_timestamp_field) {
        // When using LogNamespace::Vector always set the ingest_timestamp
        (LogNamespace::Vector, _) => {
            log.metadata_mut()
                .value_mut()
                .insert(path!("vector", "ingest_timestamp"), Utc::now());
        }
        // When LogNamespace::Legacy, only set when the option is configured
        (LogNamespace::Legacy, Some(ingestion_timestamp_field)) => {
            log.try_insert((PathPrefix::Event, ingestion_timestamp_field), Utc::now())
        }
        (LogNamespace::Legacy, None) => (),
    };
```

I originally had this function not accept a `DateTime`, and just used `now()` but wanted to limit the amount of changes throughout the sources. I think we want to be able to call `now()` for a batch of events, for efficiency. I'd like to get us to a single input, but maybe it's not possible until `Legacy` is removed?

Reading the feedback the signature would need to look like:

```rust
pub fn insert_standard_vector_source_metadata(
        &self,
        log: &mut LogEvent,
        source_name: &'static str,
        source_timestamp: DateTime<Utc>,
        ingest_timestamp: DateTime<Utc>,
)
```

Reply:
I should be the time Vector "ingested" it, but I don't care if it varies by a few nanoseconds here and there. As in, as long as it's any time while Vector is processing the event I think it's fine. Using the same timestamp for multiple events that are in a batch is fine. But don't use a value derived from the event itself.

Reply:
I've kept the function as-is for now, and added handling to `amqp` and `aws_s3` to correctly handle legacy and vector behavior (in terms of ingest time vs time from the event). The matching code can be pulled into it's own function if it pattern holds for additional sources, and I'm probably missing some ways to make it more concise.